flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/DefaultExecutionGraph.java

```diff
@@ -1031,34 +1028,18 @@ public void suspend(Throwable suspensionCause) {
             return;
         } else if (transitionState(state, JobStatus.SUSPENDED, suspensionCause)) {
             initFailureCause(suspensionCause, System.currentTimeMillis());
+            resetExecutionGraph(ExecutionJobVertex::suspend)
+                    .whenComplete(
+                            (Void ignored, Throwable throwable) -> {
+                                if (throwable != null) {
+                                    LOG.debug(
+                                            "Could not properly suspend the execution graph.",
+                                            throwable);
+                                }
 
-            incrementRestarts();
-
-            // cancel ongoing scheduling action
-            if (schedulingFuture != null) {
-                schedulingFuture.cancel(false);
-            }
-            final ArrayList<CompletableFuture<Void>> executionJobVertexTerminationFutures =
-                    new ArrayList<>(verticesInCreationOrder.size());
-
-            for (ExecutionJobVertex ejv : verticesInCreationOrder) {
-                executionJobVertexTerminationFutures.add(ejv.suspend());
-            }
-
-            final ConjunctFuture<Void> jobVerticesTerminationFuture =
-                    FutureUtils.waitForAll(executionJobVertexTerminationFutures);
-
-            checkState(jobVerticesTerminationFuture.isDone(), "Suspend needs to happen atomically");
```

Comment:
So where is this line now?

Reply:
Wait, isn't that obsolete? I deliberately removed it because the `waitForAll` has it already covered? :thinking: to be fair, we don't have a unit test to cover the class.

Reply:
thinking about it once more: It does change the behavior because the old implementation would result in a fatal error. The new version doesn't do this. I will add the precondition back. :innocent:

Reply:
waitForAll returns a future that completes _at some point_.

This check here asserted that it completed _immediately_.

Reply:
:facepalm: I was tricked by the name into assuming asynchronous execution. I fixed it